nflow-engine/src/main/java/io/nflow/engine/internal/dao/WorkflowInstanceDao.java

```diff
@@ -736,13 +736,14 @@ private long getMaxActions(Long maxActions) {
     String sql;
     if (actions < maxActions) {
       sql = "select * from nflow_workflow_state where workflow_id = ? order by action_id, state_key asc";
+      return jdbc.query(sql, new WorkflowActionStateRowMapper(), instance.id);
     } else {
       sql = "select nflow_workflow_state.* from ("
           + sqlVariants.limit("select id from nflow_workflow_action where workflow_id = ? order by id desc", maxActions)
-          + ") action_id inner join nflow_workflow_state on action_id.id = nflow_workflow_state.action_id "
+          + ") action_id inner join nflow_workflow_state on nflow_workflow_state.workflow_id = ? and action_id.id = nflow_workflow_state.action_id "
           + "order by nflow_workflow_state.action_id, nflow_workflow_state.state_key asc";
+      return jdbc.query(sql, new WorkflowActionStateRowMapper(), instance.id, instance.id);
```

Comment:
if this query is fast enough, can we use just this query and get rid of if-else to make it simpler.